### DISCUSSION TITLE
Milvus memory requirement for our case

### DISCUSSION THREAD

**Discussion by TheRabidWolverine** [milvus-io/milvus#26862] [Q&A and General discussion]
Hi, we are working on an image search engine for our product, allowing employees to compare images based on similarity, and also search by keywords ("text to image"), kind of influenced by OpenAI's CLIP paper. For that, we have to convert images into vector embeddings and store them in vector DB like Milvus, and then search for images at runtime based on query or input image. 

Questions before we start exploring, just so that we have some preliminary idea:

1. The DB has to be loaded in memory all the time, right? Like an in-memory DB like Redis? 
2. Assuming a 1024 dimension vector for each image, [this](https://github.com/milvus-io/milvus/discussions/16484#discussioncomment-6546253) link says that a 500k embedding DB size will occupy around 2 GB memory. Does that mean if the vector dimensions are say 512, will that make it around 1 GB? Is this a reasonable approximation? In the same vein, can we extrapolate that 1024 dimensioned vector, if say are 1 billion in number, will require 4000 GB memory? That is not possible in one machine, so it will need to be split into multiple machines, so is such a solution possible in Milvus, where the data is split into multiple DBs in multiple machines and the query is able to find matching images across all those DBs, sort them by relevance and return them in near real time?
3. For CLIP purpose - image search, where the queries can be as generic as possible - from "red dress" to "michael jordan slam dunk" to "nina agdal hot" to "dinosaur" to "movie poster with animals in it" to "tress with yellow flowers" - does a 1024 dimension vector sound reasonable to represent an image? Or does it require more? Or less? Of course I know this depends on dataset and fine tuning, I am just asking for some estimation.
4. The earlier referred link says that the 2 GB memory requirement for 500k 1024 dimensioned vectors is in case of **IVF_FLAT/HNSW** index, wheres the same for **IVF_PQ/IVF_SQ8** index is 600 MB. What is the difference between the two? Which one will suit our purpose better?

**Comment by yhmo** [comment]
> 1. The DB has to be loaded in memory all the time, right? Like an in-memory DB like Redis?

Yes, Milvus is an in-memory search database. Call `collection.load()` to load a single collection or `collection.release()` to release the collection from memory.


> 2. Assuming a 1024 dimension vector for each image, [this](https://github.com/milvus-io/milvus/discussions/16484#discussioncomment-6546253) link says that a 500k embedding DB size will occupy around 2 GB memory. Does that mean if the vector dimensions are say 512, will that make it around 1 GB? Is this a reasonable approximation? In the same vein, can we extrapolate that 1024 dimensioned vector, if say are 1 billion in number, will require 4000 GB memory? That is not possible in one machine, so it will need to be split into multiple machines, so is such a solution possible in Milvus, where the data is split into multiple DBs in multiple machines and the query is able to find matching images across all those DBs, sort them by relevance and return them in near real time?

Memory usage mainly depends on these factors: vector dimension, number of entities, index type, replica_number
Each dimension value is a float32 type, occupies 4 bytes. If the dimension is 512, each vector is 2048 bytes. 500k vectors will occupy 1GB in disk. But it not always load the original vectors into memory, if the 500k vectors are fully indexed by IVF_SQ8, the index size is about 250MB.
Milvus is a distributed system: https://milvus.io/docs/architecture_overview.md. The worker nodes are highly scalable. It can handle billions of data if there are sufficient query nodes and resources. The data is split into segments. By default, each segment size is 512MB: https://milvus.io/docs/data_processing.md. With proper index parameters and sufficient cpu resources, it can quickly return results(in a few seconds) for billions of vectors.

> 3. For CLIP purpose - image search, where the queries can be as generic as possible - from "red dress" to "michael jordan slam dunk" to "nina agdal hot" to "dinosaur" to "movie poster with animals in it" to "tress with yellow flowers" - does a 1024 dimension vector sound reasonable to represent an image? Or does it require more? Or less? Of course I know this depends on dataset and fine tuning, I am just asking for some estimation.


Vector ANN search can help to find out similar images. But it cannot tell you which image is "red dress". You might need some other tools to add tags for the images. Milvus can store attributes along with vector, and do hybrid search with the attributes.


> 4. The earlier referred link says that the 2 GB memory requirement for 500k 1024 dimensioned vectors is in case of **IVF_FLAT/HNSW** index, wheres the same for **IVF_PQ/IVF_SQ8** index is 600 MB. What is the difference between the two? Which one will suit our purpose better?

We provide a sizing tool to help estimate the memory usage: https://milvus.io/tools/sizing
Typically, IVF_FLAT/HNSW index is a bit larger than the original data size. IVF_PQ/IVF_SQ8 is about 25%~30% of the original data size.
HNSW is fast but requires more memory. If you want to save memory, use the IVF_SQ8. If you don't care about the recall rate, use the IVF_PQ.

**Reply by TheRabidWolverine** [reply]
Wow. That is a very detailed answer! Thanks a lot!

1. Red also will be part of the vector, since there will be enough images with red dress in it and an accompanying caption like "image of a red dress", so that means redness becomes an embedding, and will be useful to get other red dresses without caption. Is that a fair assumption?

2. When you say "IVF_PQ/IVF_SQ8 is about 25-30% of original data size", what do you mean? Original data is a list of millions of vectors, what do you mean by size of that? 

3. You said "If you want to save memory, use the IVF_SQ8". How much latency will it bring? Assuming a cluster of multiple Milvus nodes with a total of billions of vectors, will it manage to give a performance like say one gets on Google or Bing Image Search page?

4. I just searched the tool for some metrics. I tried 10 billion vectors (10000 million), it suggests 308 query nodes of 128 GB RAM each with 16 cores for HNSW, while 112 for IVF_SQ8. Isn't that an abnormally high number to store 10 billion vectors? Or is it expected?

**Reply by xiaofan-luan** [reply]
IVF_SQ8 will not reduce your vector size and it will reduce your search latency.
But you pay for recall. because of SQ8 is actually to put a float into a int8 for search, so you may lose some search accuracy

**Reply by TheRabidWolverine** [reply]
Thanks. It is likely to miss accuracy in what way? Like some vectors which are really close and similar will go missing? Or will some irrelevant vectors may also show up as nearby vectors?

**Reply by qu0laz** [reply]
@TheRabidWolverine Quick side note regarding memory size on one machine. The supermicro SuperServer 8048B-TR4FT has 96 dimms and the mem2 version supports DDR4. I just purchased one of these on ebay for $1,500 and 2tb of DDR4 (32x64gb) for $1,440. This gives me the ability to expand up to 96 x 64gb dimms which means for an extra $3k I can add 4tb of memory; total cost for a machine with 6tb of memory is under $6,000. To get a machine with 6tb of memory using 128gb dimms will be around $15k and a machine with 256gb dimms will be at least $20k. Just wanted to let you know it is possible to get up to 6tb of memory in one machine without it costing more than a small car.

**Reply by TheRabidWolverine** [reply]
Hi, just one single machine has 6 TB of memory? Wow. Is it feasible for one machine alone to host or accommodate such insanely high amount of memory? I was under the impression that large memory systems typically place several nodes in series so that the sum total of the memories of all those machines add up to the required value.

Anyway, the 6 TB machine has tobe hosted on some rackspace on some datacenter I guess, so that would incur insanely high amount of running charge per month, don't you think?
